sql3/test/defs/defs_groupby.go

```diff
@@ -257,3 +257,171 @@ var groupByTests = TableTest{
 		},
 	},
 }
+
+// groupby/distinct with sets tests
+var groupBySetDistinctTests = TableTest{
+	Table: tbl(
+		"groupby_set_test",
+		srcHdrs(
+			srcHdr("_id", fldTypeID),
+			srcHdr("ids1", fldTypeIDSet),
+			srcHdr("ss1", fldTypeStringSet),
+		),
+		srcRows(
+			srcRow(int64(1), []int64{1, 2}, []string{"a", "b"}),
+			srcRow(int64(2), []int64{3, 4}, []string{"d", "e"}),
+			srcRow(int64(3), []int64{1, 4}, []string{"a", "d"}),
+			srcRow(int64(4), []int64{3, 2}, []string{"c", "b"}),
+		),
+	),
+	SQLTests: []SQLTest{
+		{
+			SQLs: sqls(
+				"select distinct ids1 from groupby_set_test with (flatter(foo))",
+			),
+			ExpErr: "unknown query hint 'flatter'",
+		},
+		{
+			SQLs: sqls(
+				"select distinct ids1 from groupby_set_test with (flatten(foo))",
+			),
+			ExpErr: "column 'foo' not found",
+		},
+		{
+			SQLs: sqls(
+				"select distinct ids1 from groupby_set_test with (flatten(foo, bar))",
+			),
+			ExpErr: "query hint 'flatten' expected 1 parameter(s) (column name), got 2 parameters",
+		},
+		{
+			SQLs: sqls(
+				"select distinct ids1 from groupby_set_test",
+			),
+			ExpHdrs: hdrs(
+				hdr("ids1", fldTypeIDSet),
+			),
+			ExpRows: rows(
+				row([]int64{1, 2}),
+				row([]int64{3, 4}),
+				row([]int64{1, 4}),
+				row([]int64{2, 3}),
+			),
+			Compare: CompareExactUnordered,
+		},
+		{
+			SQLs: sqls(
+				"select distinct ids1 from groupby_set_test with (flatten(ids1))",
+			),
+			ExpHdrs: hdrs(
+				hdr("ids1", fldTypeIDSet),
+			),
+			ExpRows: rows(
+				row([]int64{1}),
+				row([]int64{2}),
+				row([]int64{3}),
+				row([]int64{4}),
+			),
+			Compare: CompareExactUnordered,
+		},
+		{
+			SQLs: sqls(
+				"select count(*), ids1 from groupby_set_test group by ids1",
+			),
+			ExpHdrs: hdrs(
+				hdr("", fldTypeInt),
+				hdr("ids1", fldTypeIDSet),
+			),
+			ExpRows: rows(
+				row(int64(1), []int64{1, 2}),
+				row(int64(1), []int64{3, 4}),
+				row(int64(1), []int64{1, 4}),
+				row(int64(1), []int64{2, 3}),
+			),
+			Compare: CompareExactUnordered,
+		},
+		{
+			SQLs: sqls(
+				"select count(*), ids1 from groupby_set_test with (flatten(ids1)) group by ids1",
+			),
+			ExpHdrs: hdrs(
+				hdr("", fldTypeInt),
+				hdr("ids1", fldTypeIDSet),
+			),
+			ExpRows: rows(
+				row(int64(2), []int64{1}),
+				row(int64(2), []int64{2}),
+				row(int64(2), []int64{3}),
+				row(int64(2), []int64{4}),
+			),
+			Compare: CompareExactUnordered,
+		},
+		{
+			SQLs: sqls(
+				"select distinct ss1 from groupby_set_test",
+			),
+			ExpHdrs: hdrs(
+				hdr("ss1", fldTypeStringSet),
+			),
+			ExpRows: rows(
+				row([]string{"a", "b"}),
+				row([]string{"d", "e"}),
+				row([]string{"a", "d"}),
+				row([]string{"b", "c"}),
+			),
+			Compare:        CompareExactUnordered,
+			SortStringKeys: true,
+		},
+		{
+			SQLs: sqls(
+				"select distinct ss1 from groupby_set_test with (flatten(ss1))",
```

Comment:
can we also do `select distinct ss1, ids1`? how does that interact with flattening?

i'm not sure what i expect but whatever it is, i'd like to be sure it works for all three of:
* neither flattened
* both flattened
* only one flattened

because that feels like it's a case where things might blow up.

Reply:
In this case - flatten is ignored, since we have more than on column ref.